### PR TITLE
Add container testing

### PR DESCRIPTION
**Pull request by haideriqbal**
You will need the latest data and pg-dump to make sure all tests are passing. Also, you need to create suggesters for some tests. Otherwise, follow the SOP and it will work smoothly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,6 +141,162 @@ Remember to update the file and any new experiments added to the `filesystem` di
...
+1. Clean the build directory
+2. Compile the test classes
+3. Run unit tests
+4. Run integration tests
+5. Run end-to-end tests
+6. Generate JaCoCo reports
+
+Bring it up like this (the Postgres variables can take any values, remember that the container will be removed):
+```bash
+ATLAS_DATA_PATH=/path/to/your/scxa/data \
```
Comment (alfonsomunozpomer) [inline README.md]: A wild `scxa` escaped here...

File: README.md [hunk 2]
```diff
@@ -139,6 +141,162 @@ Remember to update the file and any new experiments added to the `filesystem` di
...
+Press Ctrl+C to stop the container and clean any leftovers:
+```bash
+docker stop gxa-solrcloud-1 gxa-solrcloud-2 gxa-zk-1 gxa-zk-2 gxa-zk-3 gxa-postgres-test gxa-flyway-test &&
+docker rm gxa-solrcloud-1 gxa-solrcloud-2 gxa-zk-1 gxa-zk-2 gxa-zk-3 gxa-postgres-test gxa-flyway-test gxa-gradle
+```
+Another way of removing all stopped containers in one line you can use following statement:
+
+```bash
+docker container prune
+```
```
Comment (alfonsomunozpomer) [inline README.md]: I strongly advise against this. You might have other stopped containers unrelated to the project (I usually do) that I don’t want to remove.

File: README.md [hunk 3]
```diff
@@ -139,6 +141,162 @@ Remember to update the file and any new experiments added to the `filesystem` di
...
+
+
+If you prefer, here’s a `docker-compose run` command to execute the tests:
+```bash
+ATLAS_DATA_PATH=/path/to/your/scxa/data \
```
Comment (alfonsomunozpomer) [inline README.md]: Another wild `scxa` here!

File: README.md [hunk 4]
```diff
@@ -139,6 +141,162 @@ Remember to update the file and any new experiments added to the `filesystem` di
...
+POSTGRES_HOST=gxa-postgres-test \
+POSTGRES_DB=gxpgxatest \
+POSTGRES_USER=gxa \
+POSTGRES_PASSWORD=gxa \
+docker-compose \
+-f docker-compose-postgres-test.yml \
+-f docker-compose-solrcloud.yml \
+-f docker-compose-gradle.yml \
+run --rm --service-ports \
+scxa-gradle bash -c '
```
Comment (alfonsomunozpomer) [inline README.md]: And here...

File: README.md [hunk 5]
```diff
@@ -139,6 +141,162 @@ Remember to update the file and any new experiments added to the `filesystem` di
...
+
+In either case you may find all reports at `app/build/reports`.
+
+### Running a single test
+Many times you will find yourself working in a specific test case or class. Running all tests in such cases is
+impractical. In such situations you can use
+[Gradle’s continuous build execution](https://blog.gradle.org/introducing-continuous-build). See the example below for
+e.g. `SitemapDaoIT.java`:
+```bash
+ATLAS_DATA_PATH=/path/to/your/scxa/data \
```
Comment (alfonsomunozpomer) [inline README.md]: :eyes:

File: README.md [hunk 6]
```diff
@@ -139,6 +141,162 @@ Remember to update the file and any new experiments added to the `filesystem` di
...
+POSTGRES_HOST=gxa-postgres-test \
+POSTGRES_DB=gxpgxatest \
+POSTGRES_USER=gxa \
+POSTGRES_PASSWORD=gxa \
+docker-compose \
+-f docker-compose-postgres-test.yml \
+-f docker-compose-solrcloud.yml \
+-f docker-compose-gradle.yml \
+run --rm --service-ports \
+scxa-gradle bash -c '
```
Comment (alfonsomunozpomer) [inline README.md]: :eyes:

File: app/src/main/java/uk/ac/ebi/atlas/download/DownloadController.java [hunk 7]
```diff
@@ -43,6 +43,7 @@ public String getExperimentsListParameters(@RequestParam(defaultValue = "ftp.ebi
         try {
             ftpClient.connect(ftpHost);
             ftpClient.login("ftp", "anonymous");
+            ftpClient.enterLocalPassiveMode();
```
Comment (alfonsomunozpomer) [inline app/src/main/java/uk/ac/ebi/atlas/download/DownloadController.java:46]: Interesting...

File: app/src/test/java/uk/ac/ebi/atlas/download/DownloadControllerWIT.java [hunk 8]
```diff
@@ -20,10 +21,12 @@
 @ExtendWith(SpringExtension.class)
 @WebAppConfiguration
 @ContextConfiguration(classes = TestConfig.class)
-class DownloadControllerTest {
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class DownloadControllerWIT {
```
Comment (alfonsomunozpomer) [inline app/src/test/java/uk/ac/ebi/atlas/download/DownloadControllerWIT.java:25]: :+1:

File: docker/docker-compose-postgres-test.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
+version: "3.6"
+
+services:
+  gxa-postgres-test:
+    image: postgres:10-alpine
+    container_name: $POSTGRES_HOST
+    networks:
+      - gxa
+    restart: always
+    command: -c max_wal_size=2GB
```
Comment (alfonsomunozpomer) [inline docker/docker-compose-postgres-test.yml]: Do we need this? For fixtures  WAL shouldn’t be an issue since they are small transactions.

**Review by alfonsomunozpomer — CHANGES_REQUESTED** [review]
A few minor edits to address. Other than that all good.

**Review by lingyun1010 — COMMENTED** [review]
I cannot comment on the lines those not your commits, but please look into README file, like line 43, there are some staled instructions from scxa.

**Review by upendrakumbham — APPROVED** [review]
Remove this line 'git clone --recurse-submodules https://github.com/ebi-gene-expression-group/atlas-web-core.git && \' from the Readme as core project became submodule now. All looks good to me.

**Review by alfonsomunozpomer — APPROVED** [review]
I couldn’t spot any further issues. To me everything looks good and ready to merge.

File: README.md [hunk 10]
```diff
@@ -40,7 +44,7 @@ Notice that either way `ATLAS_DATA_PATH` will be created for you if the director
 Besides `ATLAS_DATA_PATH` you need to set some variables for the Postgres container. Use the settings below and replace
 `ATLAS_DATA_PATH` value to the directory you set up in the first step.
 
-In the `atlas-web-single-cell/docker` directory run the following:
+In the `atlas-web-bulk-cell/docker` directory run the following:
```
Comment (lingyun1010) [inline README.md:47]: The following bash code is not working for the new docker compose version anymore, please modify it to the latest command.

**Review by lingyun1010 — APPROVED** [review]
All good for me!

**Pull request merged**